File: backend/src/main/java/com/vanter/ember/printing/service/ReceiptBusinessInfo.java
package com.vanter.ember.printing.service;

import com.vanter.ember.settings.model.SettingsPayload;
import com.vanter.ember.settings.model.SettingsPayload.BusinessHoursSettings.DaySchedule;
import java.time.DayOfWeek;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * The business-info block printed under the receipt header: opening hours and, optionally, RUC,
 * address and phone. Pure — takes settings, returns plain ASCII lines (no accents: thermal
 * printers print whatever code page they default to), each wrapped and centered by
 * {@link ReceiptLayout}.
 */
final class ReceiptBusinessInfo {

    private static final Map<DayOfWeek, String> DAY_ABBR = Map.of(
            DayOfWeek.MONDAY, "Lun", DayOfWeek.TUESDAY, "Mar", DayOfWeek.WEDNESDAY, "Mie",
            DayOfWeek.THURSDAY, "Jue", DayOfWeek.FRIDAY, "Vie", DayOfWeek.SATURDAY, "Sab",
            DayOfWeek.SUNDAY, "Dom");

    private ReceiptBusinessInfo() {}

    static List<String> lines(SettingsPayload settings) {
        List<String> lines = new ArrayList<>();
        SettingsPayload.TicketSettings ticket = settings.getTicket();
        SettingsPayload.BrandingSettings branding = settings.getBranding();

        if (ticket.isShowBusinessInfo()) {
            addIfPresent(lines, "RUC: ", branding.getRuc());
            addIfPresent(lines, "", branding.getAddress());
            addIfPresent(lines, "Tel: ", branding.getPhone());
        }
        if (ticket.isShowBusinessHours()) {
            lines.addAll(hoursLines(settings));
        }
        return lines;
    }

    private static void addIfPresent(List<String> lines, String prefix, String value) {
        if (value != null && !value.isBlank()) {
            lines.add(prefix + value.strip());
        }
    }

    /**
     * The per-day schedule collapsed into runs of consecutive days with identical hours
     * ("Lun-Vie 12:00-23:00"); falls back to the single opening/closing range from Branding when
     * no per-day schedule was ever saved; empty when neither exists.
     */
    static List<String> hoursLines(SettingsPayload settings) {
        List<String> groups = scheduleGroups(settings.getBusinessHours().getSchedule());
        if (groups.isEmpty()) {
            String open = settings.getBranding().getOpeningTime();
            String close = settings.getBranding().getClosingTime();
            if (open != null && !open.isBlank() && close != null && !close.isBlank()) {
                return List.of("Horario: " + open.strip() + " - " + close.strip());
            }
            return List.of();
        }
        List<String> lines = new ArrayList<>();
        if (groups.size() == 1) {
            lines.add("Horario: " + groups.get(0));
        } else {
            lines.add("Horario:");
            lines.addAll(groups);
        }
        return lines;
    }

    private static List<String> scheduleGroups(List<DaySchedule> schedule) {
        List<String> groups = new ArrayList<>();
        if (schedule == null || schedule.isEmpty()) {
            return groups;
        }
        String runSignature = null;
        DayOfWeek runStart = null;
        DayOfWeek runEnd = null;
        for (DayOfWeek day : DayOfWeek.values()) {
            String signature = schedule.stream()
                    .filter(s -> s.getDay() == day)
                    .findFirst()
                    .map(ReceiptBusinessInfo::signature)
                    .orElse(null);
            if (signature != null && signature.equals(runSignature)) {
                runEnd = day;
                continue;
            }
            flush(groups, runSignature, runStart, runEnd);
            runSignature = signature;
            runStart = signature == null ? null : day;
            runEnd = runStart;
        }
        flush(groups, runSignature, runStart, runEnd);
        return groups;
    }

    private static String signature(DaySchedule s) {
        if (s.isClosed()) {
            return "cerrado";
        }
        if (s.getOpenTime() == null || s.getOpenTime().isBlank()
                || s.getCloseTime() == null || s.getCloseTime().isBlank()) {
            return null;
        }
        return s.getOpenTime().strip() + "-" + s.getCloseTime().strip();
    }

    private static void flush(List<String> groups, String signature, DayOfWeek start, DayOfWeek end) {
        if (signature == null || start == null) {
            return;
        }
        String label = start == end ? DAY_ABBR.get(start) : DAY_ABBR.get(start) + "-" + DAY_ABBR.get(end);
        groups.add(label + " " + signature);
    }
}

File: frontend/src/lib/receiptBusinessInfo.ts
import type { components } from '@/lib/backend-types'

type SettingsPayload = components['schemas']['SettingsPayload']
type DaySchedule = components['schemas']['DaySchedule']
type Day = NonNullable<DaySchedule['day']>

const DAYS: Day[] = ['MONDAY', 'TUESDAY', 'WEDNESDAY', 'THURSDAY', 'FRIDAY', 'SATURDAY', 'SUNDAY']
// ASCII on purpose: it mirrors what the thermal printer prints (see the backend's ReceiptBusinessInfo).
const ABBR: Record<Day, string> = {
  MONDAY: 'Lun',
  TUESDAY: 'Mar',
  WEDNESDAY: 'Mie',
  THURSDAY: 'Jue',
  FRIDAY: 'Vie',
  SATURDAY: 'Sab',
  SUNDAY: 'Dom',
}

const signature = (s: DaySchedule): string | null => {
  if (s.closed) return 'cerrado'
  if (!s.openTime?.trim() || !s.closeTime?.trim()) return null
  return `${s.openTime.trim()}-${s.closeTime.trim()}`
}

const scheduleGroups = (schedule: DaySchedule[]): string[] => {
  const groups: string[] = []
  let runSignature: string | null = null
  let runStart: Day | null = null
  let runEnd: Day | null = null
  const flush = () => {
    if (runSignature && runStart && runEnd) {
      const label = runStart === runEnd ? ABBR[runStart] : `${ABBR[runStart]}-${ABBR[runEnd]}`
      groups.push(`${label} ${runSignature}`)
    }
  }
  for (const day of DAYS) {
    const entry = schedule.find((s) => s.day === day)
    const sig = entry ? signature(entry) : null
    if (sig !== null && sig === runSignature) {
      runEnd = day
      continue
    }
    flush()
    runSignature = sig
    runStart = sig === null ? null : day
    runEnd = runStart
  }
  flush()
  return groups
}

/** Opening hours as the receipt prints them: runs of identical consecutive days, else the Branding range. */
export const hoursLines = (settings?: SettingsPayload): string[] => {
  const groups = scheduleGroups(settings?.businessHours?.schedule ?? [])
  if (groups.length === 0) {
    const open = settings?.branding?.openingTime?.trim()
    const close = settings?.branding?.closingTime?.trim()
    return open && close ? [`Horario: ${open} - ${close}`] : []
  }
  return groups.length === 1 ? [`Horario: ${groups[0]}`] : ['Horario:', ...groups]
}

/** The block under the receipt header: RUC/address/phone (opt-in) then the opening hours. */
export const businessInfoLines = (
  settings: SettingsPayload | undefined,
  opts: { showHours: boolean; showInfo: boolean }
): string[] => {
  const lines: string[] = []
  if (opts.showInfo) {
    const branding = settings?.branding
    if (branding?.ruc?.trim()) lines.push(`RUC: ${branding.ruc.trim()}`)
    if (branding?.address?.trim()) lines.push(branding.address.trim())
    if (branding?.phone?.trim()) lines.push(`Tel: ${branding.phone.trim()}`)
  }
  if (opts.showHours) lines.push(...hoursLines(settings))
  return lines
}
